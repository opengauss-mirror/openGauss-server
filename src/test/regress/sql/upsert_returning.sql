DROP SCHEMA test_upsert_returning CASCADE;
CREATE SCHEMA test_upsert_returning;
SET CURRENT_SCHEMA TO test_upsert_returning;


CREATE TEMP TABLE foo (f1 int unique, f2 text, f3 int default 42);
CREATE TEMP TABLE foo_2 (a int, b int);
INSERT INTO foo_2 select generate_series(1, 5), generate_series(1, 2);

INSERT INTO foo (f1, f2, f3)
    VALUES (1, 'test', DEFAULT), (2, 'More', 11), (3, upper('more'), 7+9) ON DUPLICATE KEY UPDATE f3 = f3+1
    RETURNING *, f1+f3 AS sum;

SELECT * FROM foo ORDER BY f1;

with t as
(
    INSERT INTO foo SELECT f1+1, f2, f3+2 FROM foo order by f3 ON DUPLICATE KEY UPDATE f3 = f3+10 RETURNING foo.*, '1'
)
select * from t;


-- error
INSERT INTO foo SELECT f1+1, f2, f3+2 FROM foo order by f3 ON DUPLICATE KEY UPDATE f1 = f1+1 RETURNING f3, f2, f1, least(f1,f3);

INSERT INTO foo SELECT f1+1, f2, f3+2 FROM foo order by f3 ON DUPLICATE KEY UPDATE f3 = f3+1 RETURNING f3, f2, f1, least(f1,f3);

-- update nothing
INSERT INTO foo SELECT f1+1, f2, f3+2 FROM foo order by f3 ON DUPLICATE KEY UPDATE f3 = f3 RETURNING f3, f2, f1, least(f1,f3);

SELECT * FROM foo ORDER BY f1;

-- Subplans and initplans in the RETURNING list

INSERT INTO foo SELECT f1+1, f2, f3+99 FROM foo order by 1, 2, 3 ON DUPLICATE KEY UPDATE f3 =f3+10 RETURNING *, f1 - 3 IN
(SELECT b FROM foo_2) AS subplan, EXISTS(SELECT * FROM foo_2) AS initplan;

SELECT * FROM foo order by 1,2,3;

DROP SCHEMA test_upsert_returning CASCADE;