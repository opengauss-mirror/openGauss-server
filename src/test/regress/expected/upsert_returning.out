DROP SCHEMA test_upsert_returning CASCADE;
ERROR:  schema "test_upsert_returning" does not exist
CREATE SCHEMA test_upsert_returning;
CREATE SCHEMA
SET CURRENT_SCHEMA TO test_upsert_returning;
CREATE TEMP TABLE foo (f1 int unique, f2 text, f3 int default 42);
NOTICE:  CREATE TABLE / UNIQUE will create implicit index "foo_f1_key" for table "foo"
CREATE TEMP TABLE foo_2 (a int, b int);
INSERT INTO foo_2 select generate_series(1, 5), generate_series(1, 2);
INSERT INTO foo (f1, f2, f3)
    VALUES (1, 'test', DEFAULT), (2, 'More', 11), (3, upper('more'), 7+9) ON DUPLICATE KEY UPDATE f3 = f3+1
    RETURNING *, f1+f3 AS sum;
 f1 |  f2  | f3 | sum
----+------+----+-----
  1 | test | 42 |  43
  2 | More | 11 |  13
  3 | MORE | 16 |  19
(3 rows)

SELECT * FROM foo ORDER BY f1;
 f1 |  f2  | f3
----+------+----
  1 | test | 42
  2 | More | 11
  3 | MORE | 16
(3 rows)

with t as
(
    INSERT INTO foo SELECT f1+1, f2, f3+2 FROM foo order by f3 on ON DUPLICATE KEY UPDATE f3 = f3+10 RETURNING foo.*, '1'
)
select * from t;
 f1 |  f2  | f3 | ?column?
----+------+----+----------
  3 | MORE | 26 | 1
  4 | MORE | 18 | 1
  2 | More | 21 | 1
(3 rows)

-- error
INSERT INTO foo SELECT f1+1, f2, f3+2 FROM foo order by f3 ON DUPLICATE KEY UPDATE f1 = f1+1 RETURNING f3, f2, f1, least(f1,f3);
ERROR:  INSERT ON DUPLICATE KEY UPDATE don't allow update on primary key or unique key.
INSERT INTO foo SELECT f1+1, f2, f3+2 FROM foo order by f3 ON DUPLICATE KEY UPDATE f3 = f3+1 RETURNING f3, f2, f1, least(f1,f3);
 f3 |  f2  | f1 | least
----+------+----+-------
 20 | MORE |  5 |     5
 27 | MORE |  3 |     3
 19 | MORE |  4 |     4
 22 | More |  2 |     2
(4 rows)

-- update nothing
INSERT INTO foo SELECT f1+1, f2, f3+2 FROM foo order by f3 ON DUPLICATE KEY UPDATE f3 = f3 RETURNING f3, f2, f1, least(f1,f3);
 f3 |  f2  | f1 | least
----+------+----+-------
 20 | MORE |  5 |     5
 22 | MORE |  6 |     6
 27 | MORE |  3 |     3
 19 | MORE |  4 |     4
 22 | More |  2 |     2
(5 rows)

SELECT * FROM foo ORDER BY f1;
 f1 |  f2  | f3
----+------+----
  1 | test | 42
  2 | More | 22
  3 | MORE | 27
  4 | MORE | 19
  5 | MORE | 20
  6 | MORE | 22
(6 rows)

-- Subplans and initplans in the RETURNING list
INSERT INTO foo SELECT f1+1, f2, f3+99 FROM foo order by 1, 2, 3 ON DUPLICATE KEY UPDATE f3 =f3+10 RETURNING *, f1 - 3 IN
(SELECT b FROM foo_2) AS subplan, EXISTS(SELECT * FROM foo_2) AS initplan;
 f1 |  f2  | f3  | subplan | initplan
----+------+-----+---------+----------
  2 | More |  32 | f       | t
  3 | MORE |  37 | f       | t
  4 | MORE |  29 | t       | t
  5 | MORE |  30 | t       | t
  6 | MORE |  32 | f       | t
  7 | MORE | 121 | f       | t
(6 rows)

SELECT * FROM foo order by 1,2,3;
 f1 |  f2  | f3
----+------+-----
  1 | test |  42
  2 | More |  32
  3 | MORE |  37
  4 | MORE |  29
  5 | MORE |  30
  6 | MORE |  32
  7 | MORE | 121
(7 rows)

DROP SCHEMA test_upsert_returning CASCADE;